### PR TITLE
openapi

### PR DESCRIPTION
**Pull request by RafiZz**

### CODE REVIEW

File: openapi/__manifest__.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
         "web_tour",
         "web_settings_dashboard",
     ],
-    "external_dependencies": {"python": [], "bin": []},
+    "external_dependencies": {"python": ['bravado_core'], "bin": []},
```
Comment (ufaks) [inline openapi/__manifest__.py]: @RafiZz please add this lib in the misc-addons/requirements.txt file

File: openapi/controllers/main.py [hunk 2]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
         }})
 
         return result
+
+
+class OAS(http.Controller):
+
+    @http.route('/api/v1/<namespace_name>/swagger.json',
```
Comment (ufaks) [inline openapi/controllers/main.py]: Why not use /api/v1/<namespace_name>.json

File: openapi/controllers/main.py [hunk 3]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+                type='http', auth='public', csrf=False)
```
Comment (ufaks) [inline openapi/controllers/main.py]: We cannot give everyone access to a spec file.
Here should be auth='user', csrf=True(it's default)

File: openapi/controllers/main.py [hunk 4]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+    def OAS_json_spec_download(self, namespace_name, **kwargs):
+        namespace = http.request.env['openapi.namespace'].sudo().search([('name', '=', namespace_name)])
```
Comment (ufaks) [inline openapi/controllers/main.py]: The 'name' field should be unique in the model

File: openapi/controllers/main.py [hunk 5]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+        if not namespace:
+            raise werkzeug.exceptions.NotFound()
+
+        current_host = http.request.env['ir.config_parameter'].sudo().get_param('web.base.url')
+        parsed_current_host = urlparse.urlparse(current_host)
+
+        spec = {
```
Comment (ufaks) [inline openapi/controllers/main.py]: I think we have to move this part to a spec generating method in the 'openapi.namespace' model?

File: openapi/controllers/main.py [hunk 6]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+            "swagger": "2.0",
+            "info": {
+                "title": "Swagger Sample App",
+                "version": "1.0.0"
+            },
+            "host": parsed_current_host.netloc,
+            "basePath": "/v1/%s" % namespace_name,
+            "schemes": [
```
Comment (ufaks) [inline openapi/controllers/main.py]: We can use urlparse.urlparse(current_host).scheme for this

File: openapi/controllers/main.py [hunk 7]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+                "https",
+                "http"
+            ],
+        }
+
+        try:
+            spec.update(namespace.get_OAS_part())
+        except SwaggerValidationError as e:
```
Comment (ufaks) [inline openapi/controllers/main.py]: I don't understand where from this exception comes.

File: openapi/controllers/main.py [hunk 8]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+        sorted_spec = OrderedDict([
+            (property_name, spec[property_name])
+            for property_name in SWAGGER_ORDERED_MAIN_PROPERTIES
+            if spec.get(property_name)
+        ])
+
+        response_params = {
+            'headers': [('Content-Type', 'application/json')]
+        }
+        if 'download' in kwargs:
```
Comment (ufaks) [inline openapi/controllers/main.py]: @yelizariev Should we give ability to open a spec file in a browser window? Or is download only available?

File: openapi/controllers/main.py [hunk 9]
```diff
@@ -39,3 +49,77 @@ def copy(obj, **d):
...
+            response_params['direct_passthrough'] = True
+
+        return werkzeug.wrappers.Response(
+            json.dumps(sorted_spec),
+            status=200,
+            **response_params
+        )
+
+
+SWAGGER_ORDERED_MAIN_PROPERTIES = [
```
Comment (ufaks) [inline openapi/controllers/main.py]: Module globals  are usually at a beginning of a module, after imports
why you use global var for this?

File: openapi/models/openapi_namespace.py [hunk 10]
```diff
@@ -35,6 +35,8 @@ class Namespace(models.Model):
         string='Latest usage'
     )
 
+    accesses = fields.One2many('openapi.access', 'namespace_id', string='Accesses')
```
Comment (ufaks) [inline openapi/models/openapi_namespace.py]: For relational fields use following ends:
M2O - _id
O2M, M2M - _ids

File: openapi/controllers/main.py (hunk 8 above)
Comment (yelizariev) [reply openapi/controllers/main.py]: We can decide later. I'm not sure now. Maybe @blaggacao have an idea

File: openapi/controllers/main.py (hunk 2 above)
Comment (blaggacao) [reply openapi/controllers/main.py]: I've seen `swagger.json` or `swagger_version_description.json`, yet most of the time `swagger` is (still) common practice.

File: openapi/controllers/main.py (hunk 3 above)
Comment (blaggacao) [reply openapi/controllers/main.py]: Problem is, that to work with swagger UI, it needs to access this link directly. Only possible workaround, I could think of: Use odoo's of token authentication and generate `?token=...` links (not sure about the exact parameters in odoo). Similar to how one time access is granted to sales document.

File: openapi/controllers/main.py (hunk 8 above)
Comment (blaggacao) [reply openapi/controllers/main.py]: As noted above, usually you spin up a swagger UI and just copy the URL for which it should render the documentation. As noted above, I think a tokenized URL would be optimum.

I think a download option can be dropped all together, in such case. (after all, API consumers are tech-savvy people and have heard about `CRTL+C` and `CTRL+V` :wink: )

File: openapi/controllers/main.py (hunk 2 above)
Comment (blaggacao) [reply openapi/controllers/main.py]: Akin to the other comments:
Most of the time a swagger file is consumed as per it's URL, so: 
`/api/v1/<namespace_name>/swagger.json'` has it all.

**Comment by blaggacao** [comment]
@yelizariev @RafiZz Please ping me again, if you want a in depth review and testing from my side. I'm happy to do that :wink:

File: openapi/tests/test_json_spec.py [hunk 11]
```diff
@@ -1,20 +1,33 @@
 # -*- coding: utf-8 -*-
 # Copyright 2018 Ivan Yelizariev <https://it-projects.info/team/yelizariev>
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+import json
+
 from odoo.tests.common import HttpCase, PORT
+from bravado_core.spec import Spec
+from swagger_spec_validator.common import SwaggerValidationError
```
[the remark this answers is not in the document]
Comment (RafiZz) [reply openapi/tests/test_json_spec.py]: @ufaks "Spec.from_dict()" raises just this type of exception.

File: openapi/views/openapi_view.xml [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                        />
+                        <field name="api_create"/>
+                        <field name="api_read"/>
+                        <field name="api_update"/>
+                        <field name="api_delete"/>
+                        <field name="api_public_methods"/>
+                        <field name="public_methods"/>
+                        <field name="private_methods"/>
+                        <!--TODO: namespace_id.name instead str(namespace_id)-->
+                        <field name="read_one_id"
```
Comment (ufaks) [inline openapi/views/openapi_view.xml]: @yelizariev Why do we use M2O to ir.exports for the read_one_id and for read_many_id fields? I think we could use M2M to ir.model.fields (with model_id in domain) for this. Then the interface would be more clear.

File: openapi/controllers/main.py [hunk 13]
```diff
@@ -15,27 +24,77 @@ def web_settings_dashboard_data(self, **kw):
...
 
         namespace_list = [
-            copy(dummy, name='magento'),
-            copy(dummy, name='ebay'),
-            copy(dummy, name='1c'),
+            {
+                'id': n.id,
+                'name': n.name,
+                'models_count': len(n.access_ids),
+                'create_count': 10,
```
Comment (ufaks) [inline openapi/controllers/main.py]: @RafiZz Please add a TODO: replace these dummy data

File: openapi/models/openapi_namespace.py [hunk 14]
```diff
@@ -57,3 +72,36 @@ def create(self, vals):
...
+                    parsed_current_host.scheme
+                ],
+            }
+            for openapi_access in record.access_ids:
+                spec.update(openapi_access.get_OAS_part())
+
+            return spec
+
+    @api.multi
+    def get_spec_url(self):
```
Comment (ufaks) [inline openapi/models/openapi_namespace.py]: @RafiZz Why did you created the def for this?

File: openapi/views/openapi_view.xml [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+                        <field name="log_request"/>
+                        <field name="log_response"/>
+                        <field name="last_log_date" readonly="1"/>
+                        <field name="token"/>
+                        <field name="spec_url" widget="url" readonly="1"/>
+                    </group>
+                </div>
+                <notebook>
+                    <page string="Accessable models">
+                        <field name="access_ids"
```
Comment (ufaks) [inline openapi/views/openapi_view.xml]: @RafiZz The readonly should be removed from here and from other similar places. Look at an example in standard odoo modules. E.g. the sale module(sale.order and sale.order.line models)

File: openapi/views/openapi_view.xml [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+                               context="{'default_namespace_id': active_id}"
+                        >
+                            <tree>
+                                <field name="active"/>
+                                <field name="model_id"/>
+                            </tree>
+                        </field>
+                    </page>
+                    <page string="Logs">
+                        <field name="log_ids"
```
Comment (ufaks) [inline openapi/views/openapi_view.xml]: I think fields of logs should be always readonly.

File: openapi/views/openapi_view.xml [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    </record>
+
+
+    <record model="ir.ui.view" id="access_ir_exports_form_view">
+        <field name="name">openapi.access.exports.form</field>
+        <field name="model">ir.exports</field>
+        <field name="arch" type="xml">
+            <tree>
+                <field name="export_fields"
+                   attrs="{'readonly': [('id', '=', False)]}"
```
Comment (ufaks) [inline openapi/views/openapi_view.xml]: This readonly condition is also strange.

**Pull request merged**